### DISCUSSION TITLE
No EdgeTPU was detected.

### DISCUSSION THREAD

**Discussion by alfwro13** [blakeblackshear/frigate#2599] [Ask A Question] [answered]
I'm running Frigate as a HassOS addon, running as an ESXi VM on VMWare ESXi 6.7.
In ESXi the Coral USB shows up as "Global Unichip product 0x089a (Connected)"
My config is as per the documentation - single USB Coral: [(https://docs.frigate.video/configuration/detectors/)]

```
detectors:
  coral:
    type: edgetpu
    device: usb
```

However as soon as I start Frigate it throws this error message:

```
[2022-01-06 18:29:06] frigate.edgetpu                ERROR   : No EdgeTPU was detected. If you do not have a Coral device yet, you must configure CPU detectors.
Process detector:coral:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/tflite_runtime/interpreter.py", line 160, in load_delegate
    delegate = Delegate(library, options)
  File "/usr/lib/python3/dist-packages/tflite_runtime/interpreter.py", line 119, in __init__
    raise ValueError(capture.message)
ValueError
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/frigate/frigate/edgetpu.py", line 156, in run_detector
    object_detector = LocalObjectDetector(
  File "/opt/frigate/frigate/edgetpu.py", line 64, in __init__
    edge_tpu_delegate = load_delegate("libedgetpu.so.1.0", device_config)
  File "/usr/lib/python3/dist-packages/tflite_runtime/interpreter.py", line 162, in load_delegate
    raise ValueError('Failed to load delegate from {}\n{}'.format(
ValueError: Failed to load delegate from libedgetpu.so.1.0
[2022-01-06 18:29:22] frigate.watchdog               INFO    : Detection appears to have stopped. Exiting frigate...
```

How do I troubleshoot this further?

**Comment by xawen** [comment, marked as answer]
I posted this in an issue a few weeks ago, so this is just a copy paste.  I've been using this solution for a few months with no issues, but it does require more (cheap) hardware.

I spent a few days trying to get USB passthrough to work into an ESXi VM. What I found is that it won't work consistently. If you do manage to get it recognized, a reboot will probably break it again.

From what I can tell, the reason for this is that the Coral is an odd device. When you first plug it in, it will show up as a "Global Unichip" device with one device ID. After it initializes, it will change to a "Google USB Device" with a different ID. The Google USB Device is the one you want to use, but it's not visible to ESXi until it's initialized...but that can't happen until the VM is attached to it.

This short version is that, from my testing, the only way to make it work is to pass the entire USB controller into the VM. I did this by buying a cheap PCIe USB3 card for my host and dedicating the whole card to my Home Assistant VM. It's been rock solid under that configuration.

**Reply by alfwro13** [reply]
I got fed up with ESXI and moved to Proxmox VE - I don't know why I have waited for so long to do it - brilliant product.
Anyhow - I have redirected whole USB controller to my HA and everything is now working just fine (inference speed around 8

**Comment by alfwro13** [comment]
I've done some more digging and I think I can see the problem:
ESXI lsusb output:
Bus 001 Device 002: ID 1a6e:089a Global Unichip Corp.

HassOS lsusb output:
Bus 001 Device 002: ID 1a6e:089a

I believe what frigate is expecting to see is this ID: 18d1:9302

I have purchased a cheap usb-c controller and see if I can pass that through to HassOS as a whole and if that fixed the problem.
Others have suggested rewriting the VendorID - I should be able to do that in HassOS but that will probably be lost on restart.

**Reply by alfwro13** [reply]
what I have found so far that the problem is in general with linux and Coral USB accelerator
I am running linux on my PC and when I connected my Coral USB directly to my PC I got exactly the same ID: 1a6e:089a
following this documentation: https://coral.googlesource.com/webcoral/
I have cloned the repo:
`git clone --recurse-submodules https://github.com/google-coral/webcoral`
installed udev rules:
`scripts/linux_device_rules.sh install`
then setup the device by running:
`make reset`

and now when I run lsusb on my linux pc I get the correct ID: ID 18d1:9302 Google Inc.

I have teseted it also on a debian VM runnin on my ESIX 6.7 host - I have passed the Coral USB to that VM and it was reporting as 1a6e:089a
Once I have followed the above that got fixed and changed to correct google dev id.

My question is how to get that working on HassOS/Frigate?

**Reply by blakeblackshear** [reply]
Does this change persist across restarts and unplugging and replugging the device?

**Reply by alfwro13** [reply]
unfortunately, no

**Reply by fernent** [reply]
it works for me too but in the docker container I have still the same error, may be because the script should be executed inside the container ? (I already tried all the other solutions changing usb, buying dedicated usb ... the thing is, frigate was working with usb coral on rpi4 before upgrades ....)

**Comment by tehniemer** [comment]
I'm having this same issue with a M.2 Accelerator with Dual Edge TPU. It was working for the longest time, not totally sure when it failed but I noticed it today.

**Reply by mirkochip88** [reply]
Same here with a M.2 Dual Edge... Any solution?
`[2022-05-02 21:17:34] frigate.edgetpu                INFO    : Attempting to load TPU as pci
Process detector:coral_pci:
[2022-05-02 21:17:47] frigate.edgetpu                ERROR   : No EdgeTPU was detected. If you do not have a Coral device yet, you must configure CPU detectors.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/tflite_runtime/interpreter.py", line 160, in load_delegate
    delegate = Delegate(library, options)
  File "/usr/lib/python3/dist-packages/tflite_runtime/interpreter.py", line 119, in __init__
    raise ValueError(capture.message)
ValueError
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/frigate/frigate/edgetpu.py", line 136, in run_detector
    object_detector = LocalObjectDetector(
  File "/opt/frigate/frigate/edgetpu.py", line 44, in __init__
    edge_tpu_delegate = load_delegate("libedgetpu.so.1.0", device_config)
  File "/usr/lib/python3/dist-packages/tflite_runtime/interpreter.py", line 162, in load_delegate
    raise ValueError('Failed to load delegate from {}\n{}'.format(
ValueError: Failed to load delegate from libedgetpu.so.1.0`

**Reply by didi767** [reply]
any updates?

**Reply by briodan** [reply]
were you ever able to figure this out?

**Comment by PeterGrace** [comment]
Hi there,

In my quest to solve this issue with passing the TPU into frigate on my synology via portainer, I found that I could start the frigate image and run the parrot example test from the docker command line with the following scenario:
  1. run `docker run -ti --rm --entrypoint=/bin/bash --privileged -v /dev/bus/usb:/dev/bus/usb blakeblackshear/frigate:0.11.0-rc2`
  2. inside container, run
```
cd ~ \
 && apt-get -y update \
 && apt-get -y install curl \
 && mkdir test_data \
 && cd test_data \
 && curl -LO https://coral.ai/static/docs/images/parrot.jpg \
 && curl -LO https://raw.githubusercontent.com/google-coral/test_data/master/mobilenet_v2_1.0_224_inat_bird_quant_edgetpu.tflite \
 && curl -LO https://raw.githubusercontent.com/google-coral/test_data/master/inat_bird_labels.txt \
 && cd .. \
 && curl -LO https://raw.githubusercontent.com/google-coral/pycoral/master/examples/classify_image.py \
 && python3 classify_image.py \
  --model test_data/mobilenet_v2_1.0_224_inat_bird_quant_edgetpu.tflite  \
  --labels test_data/inat_bird_labels.txt \
  --input test_data/parrot.jpg
```
  3. see that EdgeTPU initializes and properly classifies the CoralAI Test Parrot:
```
/root/classify_image.py:79: DeprecationWarning: ANTIALIAS is deprecated and will be removed in Pillow 10 (2023-07-01). Use Resampling.LANCZOS instead.
  image = Image.open(args.input).convert('RGB').resize(size, Image.ANTIALIAS)
----INFERENCE TIME----
Note: The first inference on Edge TPU is slow because it includes loading the model into Edge TPU memory.
187.8ms
3.8ms
3.0ms
2.9ms
2.8ms
-------RESULTS--------
Ara macao (Scarlet Macaw): 0.76953
```

Finding this to be unusual (why would this work in the same docker image, but not in my docker-compose where I'm doing essentially the same thing?) I scrutinzed the `docker inspect` output from the 'working' instantiation and the 'failing instantiation' and came to find that in my situation (trying to run frigate, via portainer) that my docker-compose in portainer had /dev/bus/usb mounted as a **devices** entry whereas in the test docker image, it was just a simple **volume**.

I attempted to restart the frigate container but with /dev/bus/usb mounted as a volume:
```
version: "3.9"
services:
  frigate:
    container_name: frigate
    privileged: true # this may not be necessary for all setups
    restart: unless-stopped
    image: blakeblackshear/frigate:0.11.0-rc2
    shm_size: "256mb" # update for your cameras based on calculation above

    volumes:
      - /etc/localtime:/etc/localtime:ro
      - /volume1/frigate/conf/config.yml:/config/config.yml:ro
      - /volume1/frigate/media:/media/frigate
      - /dev/bus/usb:/dev/bus/usb # passes the USB Coral, needs to be modified for other versions
      - type: tmpfs # Optional: 1GB of memory, reduces SSD/SD Card wear
        target: /tmp/cache
        tmpfs:
          size: 1000000000
    ports:
      - "55000:5000"
      - "21935:1935" # RTMP feeds
    environment:
      FRIGATE_RTSP_PASSWORD: "redacted"
```

amazingly, that worked and frigate initialized the tpu:
```
[2022-09-11 20:36:09] frigate.app                    INFO    : Starting Frigate (0.11.0-c461c9e)
[2022-09-11 20:36:09] frigate.config                 WARNING : The 'retain_days' config option has been DEPRECATED and will be removed in a future version. Please use the 'days' setting under 'retain'
Starting migrations
[2022-09-11 20:36:09] peewee_migrate                 INFO    : Starting migrations
There is nothing to migrate
[2022-09-11 20:36:10] peewee_migrate                 INFO    : There is nothing to migrate
[2022-09-11 20:36:10] frigate.app                    INFO    : Output process started: 216
[2022-09-11 20:36:10] detector.coral                 INFO    : Starting detection process: 214
[2022-09-11 20:36:10] ws4py                          INFO    : Using epoll
[2022-09-11 20:36:10] frigate.app                    INFO    : Camera processor started for driveway: 218
[2022-09-11 20:36:10] frigate.app                    INFO    : Capture process started for driveway: 219
[2022-09-11 20:36:10] frigate.edgetpu                INFO    : Attempting to load TPU as usb
[2022-09-11 20:36:10] ws4py                          INFO    : Using epoll
[2022-09-11 20:36:15] frigate.edgetpu                INFO    : TPU found
```

I'm not certain if this will solve your issue, but it did for me and I wanted to share my findings.

**Reply by jeremyfry** [reply]
Thank you! This fixed it for me. The guide I followed also didn't have me set the docker container to privileged which was preventing it from picking up the TPU as well.

**Reply by fhriley** [reply]
Changing `/dev/bus/usb` from a device to a volume fixed this for me as well.

**Reply by erkexzcx** [reply]
+1 worked for me too.

**Reply by rpanachi** [reply]
I tried your method (change device to volume) but it doesn't work for me.

The device seems to be bounded on the Frigate container as the output of `lsusb` shows:
```
Bus 003 Device 001: ID 1d6b:0001 Linux 5.10.110-rockchip-rk3588 ohci_hcd Generic Platform OHCI controller
Bus 001 Device 010: ID 18d1:9302
Bus 001 Device 001: ID 1d6b:0002 Linux 5.10.110-rockchip-rk3588 ehci_hcd EHCI Host Controller
```
But it wasn't recognized as expected.

Running `lsusb` on host (Debian), it shows:

```
Bus 003 Device 001: ID 1d6b:0001 Linux Foundation 1.1 root hub
Bus 001 Device 013: ID 18d1:9302 Google Inc.
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
```

I also tried to run the container using `--privileged` but doesn't work. Still receiving the error:

```
frigate.detectors.plugins.edgetpu_tfl ERROR   : No EdgeTPU was detected. If you do not have a Coral device yet, you must configure CPU detectors.
```

I'm running Frigate with Home Assistant Supervised. Any thoughts?

**Comment by spacebares** [comment]
Tried everything here and beyond, just cant get coral to show up with esxi. 
Went with proxmox, everything just works. You can pass the whole USB port to the vm. lsusb said it was a Global Unichip Corp, ran frigate, its now a google coral and it runs fine.

**Comment by jeffbeagley** [comment]
I was able to get this to work utilizing VMWare ESXI 7.0.3 with an Ubuntu VM 20.04.4 as a Kubernetes Host (k3s).

Edge TPU is in fact the USB version and is plugged into one of the USB Ports on the back of my Dell R620

I performed the following steps;

1. Within the ESXI Host, I ran the following command to disable vmkusb (this step may not be neccessary since we'll passthrough the USB Controller anyways)

	`esxcli system module set -m=vmkusb -e=FALSE`

2. Reboot ESXI Host
3. Head into vmware and go into Host => Manage and click on Hardware
4. Under PCI Devices, Toggle passthrough for the USB Controller where your Edge TPU is connected

![Screenshot 2023-01-03 231504](https://user-images.githubusercontent.com/14323118/210488979-3b93252d-c327-4b4d-a3aa-a22a01322c2d.png)
6. Locate the VM you want to pass through and click edit, adding the newly passed through USB Controller

![Screenshot 2023-01-03 231530](https://user-images.githubusercontent.com/14323118/210488981-7f4aabc6-8cdf-454a-8cc7-7dba0bf54358.png)


After performing those steps I didn't need to do anything else, both manually triggering `classify_image.py` and frigate appear to be working


Container configuration for K8s
Logs from Frigate

	2023-01-04 05:04:48.929633328  [2023-01-04 05:04:48] frigate.app                    INFO    : Starting Frigate (0.12.0-53d39a1)
	2023-01-04 05:04:49.107860320  [2023-01-04 05:04:49] peewee_migrate                 INFO    : Starting migrations
	2023-01-04 05:04:49.132055868  [2023-01-04 05:04:49] peewee_migrate                 INFO    : There is nothing to migrate
	2023-01-04 05:04:49.139096707  [2023-01-04 05:04:49] ws4py                          INFO    : Using epoll
	2023-01-04 05:04:49.182643185  [2023-01-04 05:04:49] detector.edgetpu               INFO    : Starting detection process: 691
	2023-01-04 05:04:49.186557642  [2023-01-04 05:04:49] frigate.app                    INFO    : Output process started: 693
	2023-01-04 05:04:49.194417820  [2023-01-04 05:04:49] ws4py                          INFO    : Using epoll
	2023-01-04 05:04:52.301615855  [2023-01-04 05:04:49] frigate.detectors.plugins.edgetpu_tfl INFO    : Attempting to load TPU as usb
	2023-01-04 05:04:52.326981931  [2023-01-04 05:04:49] frigate.app                    INFO    : Camera processor started for frontporch: 697
	2023-01-04 05:04:52.327419742  [2023-01-04 05:04:49] frigate.app                    INFO    : Capture process started for frontporch: 702
	2023-01-04 05:04:52.351005988  [2023-01-04 05:04:52] frigate.detectors.plugins.edgetpu_tfl INFO    : TPU found

Logs from manual invocation of classify_image.py

	image = Image.open(args.input).convert('RGB').resize(size, Image.ANTIALIAS)
	----INFERENCE TIME----
	Note: The first inference on Edge TPU is slow because it includes loading the model into Edge TPU memory.
	103.7ms
	12.0ms
	11.8ms
	11.9ms
	11.9ms

Kubernetes configuration file

	---
	apiVersion: apps/v1
	kind: Deployment
	metadata:
	name: frigate
	namespace: intranet
	labels:
		app: frigate
	spec:
	replicas: 1
	selector:
		matchLabels:
		app: frigate
	template:
		metadata:
		labels:
			app: frigate
		spec:
		# Node Affinity to attract this Deployment's pods to a specific set of worker nodes
		affinity:
			nodeAffinity:
			requiredDuringSchedulingIgnoredDuringExecution:
				nodeSelectorTerms:
				- matchExpressions:
				- key: kubernetes.io/hostname # Replace this with corresponding worker node label's key
					operator: In
					values:
					- k3s-node-3 # Replace this with corresponding worker node label's value
		volumes:
			- name: frigate-media-storage
			persistentVolumeClaim:
				claimName: frigate-pv-claim
			- name: frigate-config
			configMap:
				name: frigate-config
			- name: tmp-cache
			emptyDir: {}
			- name: dshm
			emptyDir:
				medium: Memory
				sizeLimit: 2Gi
			- name: usb-devices
			hostPath:
				path: /dev/usb/
		containers:
		- name: frigate
			image: blakeblackshear/frigate:0.12.0-beta1
			securityContext:
			privileged: true
			capabilities:
				add:
				- SYS_RAWIO
			resources:
			limits:
				ephemeral-storage: "2Gi"
			requests:
				ephemeral-storage: "1Gi"
			volumeMounts:
			- name: frigate-media-storage
				mountPath: /media
			- name: frigate-config 
				mountPath: /config/config.yml
				subPath: config.yml
			- name: frigate-config 
				mountPath: /labelmap.txt
				subPath: labelmap.txt
			- name: tmp-cache
				mountPath: /tmp/cache
			- name: dshm
				mountPath: /dev/shm
			- mountPath: /dev/usb/
				name: usb-devices
			# Many applications running for long periods of time eventually transition to broken states, 
			# and cannot recover except by being restarted. Kubernetes provides liveness probes to detect 
			# and remedy such situations.
			livenessProbe:
			httpGet:
				path: /
				port: 5000
			initialDelaySeconds: 60
			# how often to perform this check
			periodSeconds: 10
			# Sometimes, applications are temporarily unable to serve traffic. For example, an application might 
			# need to load large data or configuration files during startup, or depend on external services after 
			# startup. In such cases, you don't want to kill the application, but you don't want to send it requests either
			readinessProbe:
			httpGet:
				path: /
				port: 5000
			initialDelaySeconds: 60
			# how often to perform this check
			periodSeconds: 5

**Reply by ozett** [reply]
very valuable information. 👍 
should run also with docker instead of kubernetes ? why kubernetes?

**Reply by jeffbeagley** [reply]
> very valuable information. 👍 should run also with docker instead of kubernetes ? why kubernetes?

Should be identical using docker. 

I prefer kubernetes as my container orchestration due to its healing nature and ease of service meshing with host based routing. It's also a lot easier to do things like CI/CD . I work with kubernetes all day professionally so it just makes sense to me.

**Reply by ozett** [reply]
cool, thanks for reply.
I'll keep that in mind with kubernetes and maybe give it a try. is something new, not always just vmware

**Reply by jeffbeagley** [reply]
> cool, thanks for reply. I'll keep that in mind with kubernetes and maybe give it a try. is something new, not always just vmware

Keep in mind that kubernetes is just a container orchestration platform, its not a hypervisor (hold for one minor exception and neat project https://www.rancher.com/products/harvester ) but kubernetes requires underlying virtual machines, so you still need a hypervisor or bare metal to provide the compute.

**Reply by jeffbeagley** [reply]
Also I spoke too soon on this solution, had been messing with it for a few days and got excited to see it work. It was stable throughout the night but when I deployed a change to my camera config its now failing to find the TPU.

```
2023-01-04 17:26:17.285004032  [2023-01-04 17:26:17] ws4py                          INFO    : Using epoll
2023-01-04 17:26:17.350619197  [2023-01-04 17:26:17] detector.edgetpu               INFO    : Starting detection process: 691
2023-01-04 17:26:17.365890551  [2023-01-04 17:26:17] frigate.app                    INFO    : Output process started: 693
2023-01-04 17:26:17.366423333  [2023-01-04 17:26:17] frigate.detectors.plugins.edgetpu_tfl INFO    : Attempting to load TPU as usb
2023-01-04 17:26:17.373541873  [2023-01-04 17:26:17] ws4py                          INFO    : Using epoll
2023-01-04 17:26:17.373848160  Process detector:edgetpu:
2023-01-04 17:26:17.374358033  [2023-01-04 17:26:17] frigate.detectors.plugins.edgetpu_tfl ERROR   : No EdgeTPU was detected. If you do not have a Coral device yet, you must configure CPU detectors.
2023-01-04 17:26:17.377696249  Traceback (most recent call last):
2023-01-04 17:26:17.377762700    File "/usr/lib/python3/dist-packages/tflite_runtime/interpreter.py", line 160, in load_delegate
2023-01-04 17:26:17.377769797      delegate = Delegate(library, options)
2023-01-04 17:26:17.377775901    File "/usr/lib/python3/dist-packages/tflite_runtime/interpreter.py", line 119, in __init__
2023-01-04 17:26:17.377785431      raise ValueError(capture.message)
2023-01-04 17:26:17.377830719  ValueError
2023-01-04 17:26:17.377835606
2023-01-04 17:26:17.377840872  During handling of the above exception, another exception occurred:
```

Physically unplugged the TPU and plugged it back in and service came back online.

**Reply by ozett** [reply]
maybe USB-reset via script helps? 
i cannot find it quick, but there was an old issue and a linux script to reset the usb-bus. instead of physically plugging.
maybe thats an option for next time?

**Reply by ozett** [reply]
found it: https://github.com/tensorflow/tensorflow/issues/32743#issuecomment-907624733

**Reply by lleonardos** [reply]
Full Restart off the computer worked for me also, thanks

**Comment by bombela** [comment]
I have the same problem. At this point it has to be a bug either in the driver/firmware/hardware for the Google TPU Coral. It will work for months without an issue. And someday, it will randomly disconnect, and then even though the proper USB device ID shows up, frigate cannot find it. A reboot of the whole machine sometimes fixes it. Sometimes a full power off of the Coral is needed. It is just plain unreliable.

I would like to point out that I think that it is very irrational of frigate to exit when the AI accelerator fails. It should continue recording in a degraded mode. If the AI accelerator fails, you are out of security footage even though everything else is functional. In my case I record everything 24/7, so degraded mode would simply mean that I do not get to enjoy the AI detection atop of it.

**Comment by Bamfax** [comment]
I had this up and running for a few month now. Plugging in another device into the box suddenly disconnected the TPU (maybe because of power), so I had to puzzle how I got it up and running in the first place.

Setup: podman, debian, tpu via usb

I got it now running again and it looks to me like this now:
- After having plugged in the TPU, it always comes up as "Bus 001 Device 006: ID 1a6e:089a Global Unichip Corp."
- frigate is running in podman.
- podman need to be able to access the TPU.
- I created udevadm rules so that both device types (Global Unichip and Google Inc) are owned and writable by group "coral", which the frigate podman user is also member of
```
SUBSYSTEMS=="usb", ATTRS{idVendor}=="1a6e", ATTRS{idProduct}=="089a", MODE="0664", GROUP="coral"
SUBSYSTEMS=="usb", ATTRS{idVendor}=="18d1", ATTRS{idProduct}=="9302", MODE="0664", GROUP="coral"
```
- when frigate first starts, it accesses the TPU with the above usb device id /dev/bus/001/006 thru the frigate config "--device /dev/bus/usb/001/006:/dev/bus/usb/001/006"
- it takes frigate a minute to access the TPU
- according to another forum thread, this first access / inference calc flips the device type around from Global Unichip to Google Inc. "it is an expected behavior. When Coral USB connected it would list as Global Unichip Corp and after first inference it would change to Google Inc" (https://github.com/google-coral/edgetpu/issues/536).
- This also changes the usb device id, usually bumps it by one here. e.g. from /dev/bus/001/006 to /dev/bus/001/007.
- That's why the second udevadm from above is needed. this changes the now new "Google Inc." usb device to also be owned and writable by group coral
- the new and bumped usb device id /dev/bus/001/007 can then be put into the frigate config (having both devices in the config did not work for me)
- restarted frigate, worked.

the thread linked above also has a script which automated that first inference access. did not try it.

good look, maybe that helps.
